src/devices/cpu/z80/z80.cpp

```diff
@@ -3233,7 +3233,9 @@ void z80_device::take_nmi()
 	if (m_after_ldair) F &= ~PF;
 #endif
 
+	m_iff2 = m_iff1;
 	m_iff1 = 0;
+
 	m_r++;
```

Comment:
OK, this looks like it’s correct – NMI should save the state of IFF1 in IFF2 before clearing IFF1.  But if MAME got this wrong, how was stuff that uses non-maskable interrupts working at all up until now?

Reply:
It's hard to say why.
My assumption is what NMI is not possible initiate from software but from hardware only (like "Magic Button") that's why it becomes rare usecase in M.A.M.E. In addition NMI behave more like "jump away", not "short call of subroutine" that's why improper return back can be problematic to connect with initiation.
I have a task on my plate which will potentially help to spill more light on it leter.